music21/metadata/__init__.py

```diff
@@ -308,48 +743,281 @@ def all(self, skipContributors=False):
+        'A dedication set via grandfathered workId'
+        >>> md.title  # the uniqueName for 'otl' abbreviation
+        'A title set via grandfathered workId abbreviation'
+        >>> md.add('description', 'A description set via md.add()')
+        >>> md.description
+        'A description set via uniqueName, A description set via md.add()'
+        '''
 
-        >>> for work_id in metadata.Metadata.workIdAbbreviationDict:
-        ...    result = metadata.Metadata.abbreviationToWorkId(work_id)
+        # __getattr__ is the call of last resort after looking for bare
+        # attributes and property methods, so __getattr__ won't be called
+        # for self.software (bare attribute), or for self.composer (property),
+        # even though we certainly could handle those here. This is why we
+        # don't have to handle .composers/librettists/lyricists or
+        # .fileFormat/filePath/fileNumber here, like we do in __setattr__
+        # (which is the only call you get if you implement it, so it has
+        # to handle everything).
+        return self._getSingularAttribute(name)
+
+    def __setattr__(self, name: str, value: t.Any):
+        '''
+        Utility attribute setter for all uniqueNames, grandfathered workIds,
+        and grandfathered workId abbreviations.
 
+        These can take any single value, and will convert to the appropriate
+        internal valueType.  They will raise ValueError for iterables (other
+        than str).
         '''
-        abbreviation = abbreviation.lower()
-        if abbreviation not in Metadata.workIdAbbreviationDict:
-            raise exceptions21.MetadataException(
-                f'no such work id: {abbreviation}')
-        return Metadata.workIdAbbreviationDict[abbreviation]
+
+        # Implementation note: __setattr__ has a very different role from
+        # __getattr__.  __getattr__ is the call of last resort after looking
+        # for bare attributes and property methods, so __getattr__ won't be
+        # called for self.software (bare attribute), or for self.composer
+        # (property).  __setattr__, on the other hand, is the _only_ call
+        # you will get.  So after checking for uniqueNames, workIds, and
+        # workId abbreviations, if we haven't done the job yet, we need to
+        # make a call to super().__setattr__ to handle all the bare attributes.
+        # Unfortunately, that doesn't handle property.setters at all, so any
+        # property.setters have to be removed, and handled explicitly here.
+
+        # Is name a uniqueName?
+        if name in properties.UNIQUE_NAME_TO_NSKEY:
+            if (value is not None
+                    and isinstance(value, t.Iterable)
+                    and not isinstance(value, str)):
+                raise ValueError(f'md.{name} can only be set to a single value; '
+                                 f'set md[{name}] to multiple values instead.')
```

Comment:
This is redundant w/ two other checks below.  Do one check first of:

```
for propSet in properties.UNIQUE_NAME_TO_NSKEY, pr... etc.:
      if (name in propSet
                     and value is not None
                     and isinstance(value, t.Iterable)
                     and not isinstance(value, str)):
                 raise ValueError(f'md.{name} can only be set to a single value; '
                                   f'set md[{name}] to multiple values instead.')
```

Reply:
Done.